graph-builder/src/registry.rs

```diff
@@ -131,36 +135,44 @@ pub fn fetch_releases(
     thread_runtime.block_on(releases)
 }
 
-// Get a stream of tags
-fn get_tags(
-    repo: String,
-    authenticated_client: &dkregistry::v2::Client,
-) -> impl Stream<Item = String, Error = Error> {
-    authenticated_client
-        // According to https://docs.docker.com/registry/spec/api/#listing-image-tags
-        // the tags should be ordered lexically but they aren't
-        .get_tags(&repo, Some(20))
+/// Fetch all tags for a repository, as a stream.
+///
+/// Tags order depends on registry implementation.
+/// According to [specs](https://docs.docker.com/registry/spec/api/#listing-image-tags),
+/// remote API should return tags in lexicographic order.
+/// However on Quay 2.9 this is not true.
+fn fetch_tags(
+    client: dkregistry::v2::Client,
+    repo: &str,
+) -> impl Stream<Item = (dkregistry::v2::Client, String), Error = Error> {
+    // Paginate results, 20 tags per page.
+    let tags_per_page = Some(20);
+
+    trace!("fetching tags for repo {}", repo);
+    client
+        .get_tags(repo, tags_per_page)
+        .map(move |tags| (client.clone(), tags))
         .map_err(|e| format_err!("{}", e))
 }
 
-fn get_manifest_and_layers(
+/// Fetch manifest for a tag, and return its layers digests.
+fn fetch_layers_digests(
+    client: dkregistry::v2::Client,
     tag: String,
-    repo: String,
-    authenticated_client: &dkregistry::v2::Client,
-) -> impl Future<Item = (String, Vec<String>), Error = failure::Error> {
-    trace!("processing: {}:{}", &repo, &tag);
-    authenticated_client
-        .has_manifest(&repo, &tag, None)
-        .join(authenticated_client.get_manifest(&repo, &tag))
+    repo: &str,
+) -> impl Future<Item = (dkregistry::v2::Client, String, Vec<String>), Error = failure::Error> {
+    trace!("fetching manifest for tag {}", &tag);
+    client
+        .has_manifest(repo, &tag, None)
+        .join(client.get_manifest(repo, &tag))
         .map_err(|e| format_err!("{}", e))
-        .and_then(|(manifest_kind, manifest)| {
-            Ok((tag, get_layer_digests(&manifest_kind, &manifest)?))
-        })
+        .and_then(move |(manifest_kind, manifest)| get_layer_digests(&manifest_kind, &manifest))
+        .map(move |digests| (client, tag, digests))
 }
 
 fn find_first_release(
-    layer_digests: Vec<String>,
     authenticated_client: dkregistry::v2::Client,
+    layer_digests: Vec<String>,
```

Comment:
Why swap here?

Reply:
To have some consistency in parameter order across functions. They all take a client, in an arbitrary local order.